dozer-sql/src/pipeline/expression/onnx_udf.rs

```diff
@@ -0,0 +1,303 @@
+    let input_values = args
+        .iter()
+        .map(|arg| arg.evaluate(record, schema))
+        .collect::<Result<Vec<_>, PipelineError>>()?;
+
+    let input_shape: Vec<usize> = session.inputs[0].dimensions().map(|d| d.unwrap()).collect();
+    let output_shape: Vec<usize> = session.outputs[0].dimensions().map(|d| d.unwrap()).collect();
+    let return_type = session.outputs[0].output_type;
+
+    match input_values[0].clone() {
+        Field::String(_) => {
+            let mut input_array = vec![];
+            for field in input_values {
+                match field {
+                    Field::String(val) => input_array.push(val),
+                    _ => return Err(InvalidValue(format!("Field {field} is incompatible input"))),
+                }
+            }
+            let array = ndarray::CowArray::from(
+                Array::from_shape_vec(input_shape, input_array)
+                    .map_err(OnnxShapeErr)?
+                    .into_dyn(),
+            );
+            let input_tensor_values =
+                vec![Value::from_array(session.allocator(), &array).map_err(OnnxOrtErr)?];
+            let outputs: Vec<Value> = session.run(input_tensor_values).map_err(OnnxOrtErr)?;
+            let output = outputs[0].borrow();
+
+            // number of output validation
+            assert_eq!(outputs.len(), 1);
+
+            match return_type {
+                TensorElementDataType::Float16 => {
+                    let output_array_view = output.try_extract::<f16>().map_err(OnnxOrtErr)?;
+                    assert_eq!(output_array_view.view().shape(), output_shape);
+                    Ok(Field::Float(OrderedFloat(output_array_view.view().deref()[0].into())))
+                }
+                TensorElementDataType::Float32 => {
+                    let output_array_view = output.try_extract::<f32>().map_err(OnnxOrtErr)?;
+                    assert_eq!(output_array_view.view().shape(), output_shape);
+                    Ok(Field::Float(OrderedFloat(output_array_view.view().deref()[0].into())))
+                }
+                TensorElementDataType::Float64 => {
+                    let output_array_view = output.try_extract::<f64>().map_err(OnnxOrtErr)?;
+                    assert_eq!(output_array_view.view().shape(), output_shape);
+                    Ok(Field::Float(OrderedFloat(output_array_view.view().deref()[0].into())))
+                }
+                _ => Err(InvalidValue(format!(
+                    "{return_type:?} is incompatible return type"
+                ))),
+            }
+        }
+        Field::UInt(_) => {
+            warn!("Precision loss happens due to conversion from u64 to u32");
+            let mut input_array = vec![];
+            for field in input_values {
+                match field {
+                    Field::UInt(val) => input_array.push(u32::try_from(val)
+                        .map_err(|_e| InvalidType(field.clone(), format!("{:?}", return_type)))?),
```

Comment:
This error is not informational. I suggest creating new errors instead of using existing ones.

Reply:
ok